src/linkExposure.js

```diff
@@ -21,246 +25,458 @@ permissions.check({
-                    if (linkMatcher.matches(result)) {
-                        linkExposure.resolvedUrl = result;
-                    }
-                }, function (error) {
-                    linkExposure.error = error.message;
-                    linkExposure.resolutionSucceded = false;
-                }).finally(async function () {
-                    if (!linkExposure.resolutionSucceded || linkExposure.resolvedUrl !== undefined)
-                        await createLinkExposureRecord(linkExposure, nextLinkExposureIdCounter);
-                });
+    // Store the listener information in a record
+    linkExposureUpdateListeners.set(listener, {
+        linkMatchPatternSet,
+        pageMatchPatternSet,
+        privateWindows,
+        contentScript
+    });
+}
+
+/**
+ * Callback for removing an onLinkExposureUpdate listener.
+ * @param {linkExposureUpdateListener} listener - The listener that is being removed.
+ * @private
+ */
+function removeUpdateListener(listener) {
+    // If the listener has a record, unregister its content script and delete
+    // the record
+    const listenerRecord = linkExposureUpdateListeners.get(listener);
+    if(listenerRecord !== undefined) {
+        listenerRecord.contentScript.unregister();
+        linkExposureUpdateListeners.delete(listener);
+    }
+}
+
+/**
+ * Callback for a link exposure update message from the content script.
+ * @param {Options} linkExposureUpdate - The update message.
+ * @param {string} linkExposureUpdate.pageId - The page ID for the page where
+ * the content script is running.
+ * @param {string} linkExposureUpdate.url - The URL, without a hash, for the page
+ * where the content script is running.
+ * @param {boolean} linkExposureUpdate.privateWindow - Whether the page where the
+ * content script is running is in a private window.
+ * @param {string[]} linkExposureUpdate.linkUrls - The links on the page that the
+ * user was exposed to.
+ */
+function messageListener({ pageId, url, privateWindow, linkUrls }) {
+    // Increment the count of pending link exposure updates for the page
+    let pendingLinkExposureCount = pendingPageLinkExposureUpdates.get(pageId);
+    pendingLinkExposureCount = pendingLinkExposureCount === undefined ? 1 : pendingLinkExposureCount + 1;
+    pendingPageLinkExposureUpdates.set(pageId, pendingLinkExposureCount);
+
+    // Resolve all the link URLs in the update, converting each URL into a
+    // Promise<string>
+    const resolvedLinkUrlPromises = linkUrls.map((linkUrl) => {
+        return linkResolution.resolveUrl(linkUrl);
+    });
+
+    // Once resolution is complete, notify the linkExposureUpdate listeners
+    Promise.allSettled(resolvedLinkUrlPromises).then(async (results) => {
+        // For each link URL, if we have a resolved URL, use that
+        // If we don't have a resolved URL, use the original URL with
+        // cache, shim, and link decoration parsing
+        for(let i = 0; i < linkUrls.length; i++) {
```

Comment:
Nit, but if `linkUrls` is an array and you need the loop counter, I think you could use `for ... of` just to simplify things a little:

```suggestion
        for(const [i, linkUrl] of linkUrls.entries()) {
```

Reply:
Not important, just nice to use iterators and not worry about off-by-one etc. (of course in this case since you're indexing it into this other `results` array it doesn't help /that/ much :)) Makes it more consistent with the use of `Map` elsewhere etc